src/phy/mod.rs

```diff
@@ -19,87 +19,81 @@
 //!
 /*!
 ```rust
-use std::slice;
-use smoltcp::{Error, Result};
-use smoltcp::phy::{DeviceCapabilities, Device};
+use smoltcp::Result;
+use smoltcp::phy::{self, DeviceCapabilities, Device};
 
-const TX_BUFFERS: [*mut u8; 2] = [0x10000000 as *mut u8, 0x10001000 as *mut u8];
-const RX_BUFFERS: [*mut u8; 2] = [0x10002000 as *mut u8, 0x10003000 as *mut u8];
-
-fn rx_full() -> bool {
-    /* platform-specific code to check if an incoming packet has arrived */
-    false
+struct StmPhy {
+    rx_buffer: [u8; 1536],
+    tx_buffer: [u8; 1536],
 }
 
-fn rx_setup(_buf: *mut u8, _length: &mut usize) {
-    /* platform-specific code to receive a packet into a buffer */
+impl<'a> StmPhy {
+    fn new() -> StmPhy {
+        StmPhy {
+            rx_buffer: [0; 1536],
+            tx_buffer: [0; 1536],
+        }
+    }
 }
 
-fn tx_empty() -> bool {
-    /* platform-specific code to check if an outgoing packet can be sent */
-    false
-}
+impl<'a> phy::Device<'a> for StmPhy {
+    type RxToken = StmPhyRxToken<'a>;
+    type TxToken = StmPhyTxToken<'a>;
 
-fn tx_setup(_buf: *const u8, _length: usize) {
-    /* platform-specific code to send a buffer with a packet */
-}
-
-# #[allow(dead_code)]
-pub struct EthernetDevice {
-    tx_next: usize,
-    rx_next: usize
-}
+    fn receive(&'a mut self) -> Option<(Self::RxToken, Self::TxToken)> {
+        Some((StmPhyRxToken(&mut self.rx_buffer[..]),
+              StmPhyTxToken(&mut self.tx_buffer[..])))
+    }
 
-impl Device for EthernetDevice {
-    type RxBuffer = &'static [u8];
-    type TxBuffer = EthernetTxBuffer;
+    fn transmit(&'a mut self) -> Option<Self::TxToken> {
+        Some(StmPhyTxToken(&mut self.tx_buffer[..]))
+    }
 
     fn capabilities(&self) -> DeviceCapabilities {
-        let mut caps = DeviceCapabilities::default();
-        caps.max_transmission_unit = 1536;
-        caps.max_burst_size = Some(2);
```

Comment:
This has only a single buffer so max_burst_size should be `Some(1)`, to cap the receive window.